### PR TITLE
MAILBOX-354 Handle registration with MailboxId

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java [hunk 1]
```diff
@@ -61,9 +63,8 @@ public void testStressTest() throws InterruptedException, MailboxException {
         MailboxSession session = mailboxManager.createSystemSession(username);
         mailboxManager.startProcessingRequest(session);
         final MailboxPath path = MailboxPath.forUser(username, "INBOX");
-        mailboxManager.createMailbox(path, session);
-        mailboxManager.addListener(path, new MailboxListener() {
-
+        Optional<MailboxId> mailboxId = mailboxManager.createMailbox(path, session);
+        mailboxManager.addListener(mailboxId.get(), new MailboxListener() {
```
Comment (trantienduchn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java]: is this safe to call directly `Optional.get()` ?

**Comment by chibenwa** [comment]
```
[37617ba65dba2b7c9ba6f6240b48967ad644aef7] [INFO] Apache James MPT Imap Mailbox - Maildir ............ FAILURE [ 47.840 s]
```

**Comment by chibenwa** [comment]
```

[37617ba65dba2b7c9ba6f6240b48967ad644aef7] org.apache.james.mailbox.exception.MailboxNotFoundException: 1 can not be found
[37617ba65dba2b7c9ba6f6240b48967ad644aef7] 	at org.apache.james.mailbox.maildir.mail.MailboxCache.getCachedMailbox(MailboxCache.java:61)
[37617ba65dba2b7c9ba6f6240b48967ad644aef7] 	at org.apache.james.mailbox.maildir.mail.MaildirMailboxMapper.findMailboxById(MaildirMailboxMapper.java:119)
[37617ba65dba2b7c9ba6f6240b48967ad644aef7] 	at org.apache.james.mailbox.store.StoreMailboxManager.getMailbox(StoreMailboxManager.java:466)
[37617ba65dba2b7c9ba6f6240b48967ad644aef7] 	at org.apache.james.imap.processor.base.SelectedMailboxImpl.getPath(SelectedMailboxImpl.java:164)
[37617ba65dba2b7c9ba6f6240b48967ad644aef7] 	at org.apache.james.imap.processor.AbstractMailboxProcessor.getSelectedMailbox(AbstractMailboxProcessor.java:414)
```

File: protocols/imap/src/main/java/org/apache/james/imap/api/process/SelectedMailbox.java [hunk 2]
```diff
@@ -89,6 +91,13 @@
      */
     MailboxPath getPath();
 
+    /**
+     * Return the mailboxId of the selected Mailbox.
+     *
+     * This is beneficial as the MailboxId is immutable.
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/process/SelectedMailbox.java]: from the perspective of the API user, this sentence doesn't make much sense IMO.

**Comment by chibenwa** [comment]
We need to handle https://issues.apache.org/jira/browse/MAILBOX-292 and give persistent mailboxId to MailDir - or disable MailDir

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
We need to use UUID as a underlying type for MailDirId

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java [hunk 3]
```diff
@@ -294,6 +301,49 @@ private void saveUidValidity(long uidValidity) throws IOException {
...
+    /**
+     * Read the uidValidity of the given mailbox from the file system.
+     * If the respective file is not yet there, it gets created and
+     * filled with a brand new uidValidity.
+     * @return The uidValidity
+     * @throws IOException if there are problems with the validity file
+     */
+    public MaildirId readMailboxId() throws IOException {
+        if (!mailboxIdFile.exists()) {
+            throw new RuntimeException("Missing " + mailboxIdFile);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java]: See https://github.com/linagora/james-project/pull/1979#commitcomment-31510882

We decided to no more throw on non-existent file to allow a smooth migration...

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMailboxMapper.java [hunk 4]
```diff
@@ -152,8 +152,11 @@ public boolean hasChildren(Mailbox mailbox, char delimiter) throws MailboxExcept
 
     @Override
     public MailboxId save(Mailbox mailbox) throws MailboxException {
+        MaildirId maildirId = Optional.ofNullable(mailbox.getMailboxId())
+            .map(mailboxId -> (MaildirId) mailboxId)
+            .orElseGet(() -> MaildirId.of(Math.abs(new Random().nextInt())));
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMailboxMapper.java]: 100% agree see https://github.com/linagora/james-project/pull/1979#issuecomment-443182703

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**